### PR TITLE
Add support for keccak_utils

### PR DESCRIPTION
**Pull request by entropidelic**
# keccak_utils

## Description
This PR implements the `keccak_utils` module of the `cairo_keccak` folder, which is needed for the hints used in it. The Python module can be seen [here](https://github.com/starkware-libs/cairo-lang/blob/13cef109cd811474de114925ee61fd5ac84a25eb/src/starkware/cairo/common/cairo_keccak/keccak_utils.py).

## Checklist
- [ ] Linked to Github Issue
- [x] Unit tests added
- [ ] Integration tests added.
- [ ] This change requires new documentation.
  - [ ] Documentation has been added/updated.

### CODE REVIEW

File: src/vm/hints/cairo_keccak/keccak_utils.rs [hunk 1]
```diff
@@ -0,0 +1,411 @@
...
+        for (y, b_row) in b.iter_mut().enumerate().take(u) {
+            b_row[(beta * x + alpha * y).mod_floor(&u)] =
+                rot_left(&a_tmp[x][y], rho_offsets[x][y], w);
+        }
+    }
+
+    for x in 0..u {
+        for y in 0..u {
+            a_tmp[x][y] =
+                &b[x][y] ^ ((!&b[(x + 1).mod_floor(&u)][y]) & &b[(x + 2).mod_floor(&u)][y]);
```
Comment (Oppen) [inline src/vm/hints/cairo_keccak/keccak_utils.rs:145]: Is this negation intended to be boolean or binary?

File: src/vm/hints/cairo_keccak/keccak_utils.rs [hunk 2]
```diff
@@ -0,0 +1,411 @@
...
+        }
+    }
+
+    a_tmp[0][0] ^= rc;
+
+    a_tmp
+}
+
+fn keccak_func(
+    values: Vec<BigInt>,
```
Comment (Oppen) [inline src/vm/hints/cairo_keccak/keccak_utils.rs:155]:
```suggestion
    values: &Vec<BigInt>,
```

File: src/vm/hints/cairo_keccak/keccak_utils.rs [hunk 3]
```diff
@@ -0,0 +1,411 @@
...
+
+    for x in 0..u {
+        let mut row = Vec::new();
+        for y in 0..u {
+            row.push(values[u * y + x].clone())
+        }
+        value_matrix.push(row);
+    }
+
+    let w = 2_usize.pow(ell);
```
Comment (Oppen) [inline src/vm/hints/cairo_keccak/keccak_utils.rs:176]: Shift.

File: src/vm/hints/cairo_keccak/keccak_utils.rs [hunk 4]
```diff
@@ -0,0 +1,411 @@
...
+    ell: usize,
+    u: usize,
+    alpha: u32,
+    beta: u32,
+    rounds: Option<usize>,
+) -> Result<Vec<u8>, KeccakError> {
+    /*
+    Computes the keccak block permutation on a u**2*2**ell-bit message (pads with zeros).
+    */
+    let w = 2_usize.pow(ell as u32);
```
Comment (Oppen) [inline src/vm/hints/cairo_keccak/keccak_utils.rs:207]: Shift.

File: src/vm/hints/cairo_keccak/keccak_utils.rs [hunk 5]
```diff
@@ -0,0 +1,411 @@
...
+
+    let as_bigint: BigInt = BigInt::from_bytes_le(Sign::Plus, &message);
+
+    if as_bigint >= bigint!(1_i32).shl(u * u * w) {
+        return Err(KeccakError::BigIntMaxSize(as_bigint));
+    }
+
+    let mut as_integers = Vec::new();
+    for i in 0..(u * u) {
+        let integer = (as_bigint.clone() >> (i * w)) & (bigint!(2_u128.pow(w as u32) - 1));
```
Comment (Oppen) [inline src/vm/hints/cairo_keccak/keccak_utils.rs:224]: Shift. Avoid the clone.

**Review by Oppen — CHANGES_REQUESTED** [review]

**Comment by Oppen** [comment]
Reminder to myself: I still need to make a more thorough review to the algorithm and vectors sizes (maybe they can be made to fit the stack).

File: src/vm/hints/cairo_keccak/keccak_utils.rs [hunk 6]
```diff
@@ -0,0 +1,411 @@
...
+    let mut xy_pairs = HashSet::new();
+    let mut offset = 0;
+    let mut result = vec![vec![0; u]; u];
+
+    for t in 1..m {
+        xy_pairs.insert((x, y));
+        offset = (offset + t).mod_floor(&w);
+        result[x][y] = offset;
+        // The official definition is (alpha, beta) = (3, 2) for u = 5. Any other u has no official
+        // definition, but the iteration must go over each (x, y) != (0, 0) pair exactly once.
```
[the remark this answers is not in the document]
Comment (entropidelic) [reply src/vm/hints/cairo_keccak/keccak_utils.rs:44]: that comment was pasted from the python version literally, so I don't really know either :-P

**Pull request closed without merging**